File: examples/go/token-ft/FungibleToken_test.go
package contract

import (
	"testing"
	"tsop"
)

var (
	alice   = tsop.PubKey("alice_pubkey_33bytes_placeholder!")
	bob     = tsop.PubKey("bob___pubkey_33bytes_placeholder!")
	tokenId = tsop.ByteString("test-token-001")
)

func newToken(owner tsop.PubKey, balance tsop.Bigint) *FungibleToken {
	return &FungibleToken{Owner: owner, Balance: balance, TokenId: tokenId}
}

func TestFungibleToken_Transfer(t *testing.T) {
	c := newToken(alice, 100)
	c.Transfer(tsop.MockSig(), bob, 30, 1000)
	out := c.Outputs()
	if len(out) != 2 {
		t.Fatalf("expected 2 outputs, got %d", len(out))
	}
	if out[0].Values[0] != bob {
		t.Error("output[0] owner should be bob")
	}
	if out[0].Values[1] != tsop.Bigint(30) {
		t.Errorf("output[0] balance: expected 30, got %v", out[0].Values[1])
	}
	if out[1].Values[1] != tsop.Bigint(70) {
		t.Errorf("output[1] balance: expected 70, got %v", out[1].Values[1])
	}
}

func TestFungibleToken_Transfer_ZeroAmount_Fails(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected assertion failure")
		}
	}()
	newToken(alice, 100).Transfer(tsop.MockSig(), bob, 0, 1000)
}

func TestFungibleToken_Send(t *testing.T) {
	c := newToken(alice, 100)
	c.Send(tsop.MockSig(), bob, 1000)
	if len(c.Outputs()) != 1 {
		t.Fatalf("expected 1 output, got %d", len(c.Outputs()))
	}
}

func TestFungibleToken_Merge(t *testing.T) {
	c := newToken(alice, 50)
	c.Merge(tsop.MockSig(), 200, 1000)
	if len(c.Outputs()) != 1 {
		t.Fatalf("expected 1 output, got %d", len(c.Outputs()))
	}
}

func TestFungibleToken_Merge_LessThanBalance_Fails(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected assertion failure")
		}
	}()
	newToken(alice, 100).Merge(tsop.MockSig(), 50, 1000)
}

func TestFungibleToken_Compile(t *testing.T) {
	if err := tsop.CompileCheck("FungibleTokenExample.tsop.go"); err != nil {
		t.Fatalf("TSOP compile check failed: %v", err)
	}
}

File: examples/rust/token-ft/FungibleToken_test.rs
// Contract logic tests for FungibleToken.
//
// The contract struct is defined inline (not via #[path]) because the
// add_output output-tracking requires fields and methods that are test
// infrastructure, not part of the TSOP contract.

use tsop::prelude::*;

#[derive(Clone)]
struct FtOutput { satoshis: Bigint, owner: PubKey, balance: Bigint }

struct FungibleToken {
    owner: PubKey,
    balance: Bigint,
    token_id: ByteString,
    outputs: Vec<FtOutput>,
}

impl FungibleToken {
    fn add_output(&mut self, satoshis: Bigint, owner: PubKey, balance: Bigint) {
        self.outputs.push(FtOutput { satoshis, owner, balance });
    }

    fn transfer(&mut self, sig: &Sig, to: PubKey, amount: Bigint, output_satoshis: Bigint) {
        assert!(check_sig(sig, &self.owner));
        assert!(amount > 0);
        assert!(amount <= self.balance);
        let change_owner = self.owner.clone();
        let change_balance = self.balance - amount;
        self.add_output(output_satoshis, to, amount);
        self.add_output(output_satoshis, change_owner, change_balance);
    }

    fn send(&mut self, sig: &Sig, to: PubKey, output_satoshis: Bigint) {
        assert!(check_sig(sig, &self.owner));
        self.add_output(output_satoshis, to, self.balance);
    }

    fn merge(&mut self, sig: &Sig, total_balance: Bigint, output_satoshis: Bigint) {
        assert!(check_sig(sig, &self.owner));
        assert!(total_balance >= self.balance);
        let owner = self.owner.clone();
        self.add_output(output_satoshis, owner, total_balance);
    }
}

fn alice() -> PubKey { b"alice_pubkey_33bytes_placeholder!".to_vec() }
fn bob() -> PubKey { b"bob___pubkey_33bytes_placeholder!".to_vec() }

fn new_token(owner: PubKey, balance: Bigint) -> FungibleToken {
    FungibleToken { owner, balance, token_id: b"test-token-001".to_vec(), outputs: vec![] }
}

#[test]
fn test_transfer() {
    let mut c = new_token(alice(), 100);
    c.transfer(&mock_sig(), bob(), 30, 1000);
    assert_eq!(c.outputs.len(), 2);
    assert_eq!(c.outputs[0].owner, bob());
    assert_eq!(c.outputs[0].balance, 30);
    assert_eq!(c.outputs[1].owner, alice());
    assert_eq!(c.outputs[1].balance, 70);
}

#[test]
#[should_panic]
fn test_transfer_zero_amount_fails() {
    new_token(alice(), 100).transfer(&mock_sig(), bob(), 0, 1000);
}

#[test]
#[should_panic]
fn test_transfer_exceeds_balance_fails() {
    new_token(alice(), 100).transfer(&mock_sig(), bob(), 101, 1000);
}

#[test]
fn test_send() {
    let mut c = new_token(alice(), 100);
    c.send(&mock_sig(), bob(), 1000);
    assert_eq!(c.outputs.len(), 1);
    assert_eq!(c.outputs[0].owner, bob());
    assert_eq!(c.outputs[0].balance, 100);
}

#[test]
fn test_merge() {
    let mut c = new_token(alice(), 50);
    c.merge(&mock_sig(), 200, 1000);
    assert_eq!(c.outputs.len(), 1);
    assert_eq!(c.outputs[0].balance, 200);
}

#[test]
#[should_panic]
fn test_merge_less_than_balance_fails() {
    new_token(alice(), 100).merge(&mock_sig(), 50, 1000);
}

#[test]
fn test_compile() {
    tsop::compile_check(
        include_str!("FungibleTokenExample.tsop.rs"),
        "FungibleTokenExample.tsop.rs",
    ).unwrap();
}
